pkg/route/controller/ingress/ingress.go

```diff
@@ -0,0 +1,548 @@
+			GenerateName: ingress.Name + "-",
+			Namespace:    ingress.Namespace,
+			Labels:       ingress.Labels,
+			Annotations:  ingress.Annotations,
+			OwnerReferences: []metav1.OwnerReference{
+				{APIVersion: "extensions/v1beta1", Kind: "Ingress", Controller: &t, Name: ingress.Name, UID: ingress.UID},
+			},
+		},
+		Spec: routev1.RouteSpec{
+			Host: rule.Host,
+			Path: path.Path,
+			To: routev1.RouteTargetReference{
+				Name: path.Backend.ServiceName,
+			},
+			Port: &routev1.RoutePort{
+				TargetPort: *targetPort,
+			},
+			TLS: tlsConfig,
+		},
+	}
+}
+
+func preserveRouteAttributesFromExisting(r, existing *routev1.Route) {
+	r.Name = existing.Name
+	r.GenerateName = ""
+	r.Spec.To.Weight = existing.Spec.To.Weight
+	if r.Spec.TLS != nil && existing.Spec.TLS != nil {
+		r.Spec.TLS.CACertificate = existing.Spec.TLS.CACertificate
+		r.Spec.TLS.DestinationCACertificate = existing.Spec.TLS.DestinationCACertificate
+		r.Spec.TLS.InsecureEdgeTerminationPolicy = existing.Spec.TLS.InsecureEdgeTerminationPolicy
+	}
+}
+
+func routeMatchesIngress(
+	route *routev1.Route,
+	ingress *extensionsv1beta1.Ingress,
+	rule *extensionsv1beta1.IngressRule,
+	path *extensionsv1beta1.HTTPIngressPath,
+	secretLister corelisters.SecretLister,
+	serviceLister corelisters.ServiceLister,
+) bool {
+	match := route.Spec.Host == rule.Host &&
+		route.Spec.Path == path.Path &&
+		route.Spec.To.Name == path.Backend.ServiceName &&
+		route.Spec.Port != nil &&
+		route.Spec.WildcardPolicy == routev1.WildcardPolicyNone &&
+		len(route.Spec.AlternateBackends) == 0
+	if !match {
+		return false
+	}
+
+	targetPort := targetPortForService(ingress.Namespace, path, serviceLister)
+	if targetPort == nil || *targetPort != route.Spec.Port.TargetPort {
+		// not valid
+		return false
+	}
+
+	var secret *v1.Secret
+	if name, ok := referencesSecret(ingress, rule.Host); ok {
+		secret, _ = secretLister.Secrets(ingress.Namespace).Get(name)
+		if secret == nil {
+			return false
+		}
+	}
+	if !secretMatchesRoute(secret, route.Spec.TLS) {
+		return false
+	}
+	return true
+}
+
+func targetPortForService(namespace string, path *extensionsv1beta1.HTTPIngressPath, serviceLister corelisters.ServiceLister) *intstr.IntOrString {
+	service, err := serviceLister.Services(namespace).Get(path.Backend.ServiceName)
+	if err != nil {
+		// service doesn't exist yet, wait
+		return nil
+	}
+	if path.Backend.ServicePort.Type == intstr.String {
+		expect := path.Backend.ServicePort.StrVal
+		for _, port := range service.Spec.Ports {
+			if port.Name == expect {
+				return &port.TargetPort
+			}
+		}
+	} else {
+		for _, port := range service.Spec.Ports {
+			expect := path.Backend.ServicePort.IntVal
+			if port.Port == expect {
+				return &port.TargetPort
+			}
+		}
+	}
+	return nil
+}
+
+func secretMatchesRoute(secret *v1.Secret, tlsConfig *routev1.TLSConfig) bool {
+	if secret == nil {
+		return tlsConfig == nil
+	}
+	if secret.Type != v1.SecretTypeTLS {
+		return tlsConfig == nil
```

Comment:
This is to avoid allocating in the common case.  This controller gets woken up on any secret change in a namespace which is expected to be common, therefore we have to be very careful not to over allocate during evaluation.

Reply:
I don't follow.  Is it common that
* an ingress exists,
* a corresponding route exists,
* the ingress references a secret,
* and that secret exists,
* but the secret is the wrong type?

If this is a case worth optimizing, would it also make sense to return true in the case that the secret does not exist (i.e., `return true` from `routeMatchesIngress` if `secretLister.Secrets(ingress.Namespace).Get(name)` is `nil`)? Why treat secret-does-not-exist differently from secret-exists-but-has-the-wrong-type (which probably means the secret does not exist, and another secret is using its name)?

Reply:
Any change to a namespace for a secret will cause us to need to reevaluate every ingress in the namespace.  This is a level driven controller, so we will run over this code path for every ingress

1. everytime a secret changes in a namespace
2. everytime a service changes in a namespace
3. everytime a resync interval hits

So it does matter to be as efficient as possible because this line of code will be invoked quite often.

The reason we return sales is because the method does what it says on the tin - you can't reference a non TLS secret type from an ingress, so returning false means "doesn't match".